app/app.go

```diff
@@ -0,0 +1,190 @@
+	if mainLoop == nil {
+		return nil, errors.New("main loop is nil")
+	}
+	return &App{
+		ctx:      ctx,
+		mainLoop: mainLoop,
+	}, nil
+}
+
+// MustNew returns a new App, but panics if there is an error.
+func MustNew(ctx context.Context, mainLoop MainLoopFunc) *App {
+	app, err := New(ctx, mainLoop)
+	if err != nil {
+		panic(err)
+	}
+	return app
+}
+
+// Shutdown calls all shutdown methods, in order they were added.
+func (a *App) Shutdown(ctx context.Context) error {
+	const op = errors.Op("app.App.Shutdown")
+	if a.ShutdownTimeout > 0 {
+		var cancel func()
+		ctx, cancel = context.WithTimeout(ctx, a.ShutdownTimeout)
+		defer cancel()
+	}
+	var err error
+	select {
+	case <-ctx.Done():
+		err = ctx.Err()
+	case err = <-a.shutdownAllHandlers(ctx):
+	}
+	if err != nil {
+		return errors.E(op, err)
+	}
+	return nil
+}
+
+func (a *App) shutdownAllHandlers(ctx context.Context) chan error {
+	const op = errors.Op("shutdownAllHandlers")
+	done := make(chan error)
+	go func() {
+		defer close(done)
+		for _, h := range a.shutdownHandlers {
+			if ctx.Err() != nil {
+				done <- errors.E(op, ctx.Err())
+			}
+			err := a.executeHandlerShutdown(ctx, h)
+			if err != nil {
+				done <- errors.E(op, err)
+			}
+		}
+	}()
+	return done
+}
+
+func (a *App) executeHandlerShutdown(ctx context.Context, h shutdownHandler) error {
+	const op = errors.Op("executeHandlerShutdown")
+
+	if h.Timeout > 0 {
+		var cancel func()
+		ctx, cancel = context.WithTimeout(ctx, h.Timeout)
+		defer cancel()
+	}
+
+	err := h.Shutdown(ctx)
+	if err != nil {
+		err = errors.E(op, errors.E(errors.Op(h.Name), err))
+		switch h.Policy {
+		case ErrorPolicyWarn:
+			log.Ctx(a.ctx).Warn().Err(err).Msg("Shutdown handler failed")
+		case ErrorPolicyAbort:
+			// No need for logging here, this will happen latter
+			return err
+		case ErrorPolicyFatal:
+			log.Ctx(a.ctx).Fatal().Err(err).Msg("Shutdown handler failed")
+		case ErrorPolicyPanic:
+			panic(err)
+		default:
+			panic(errors.Errorf("invalid error policy: %v", h.Policy))
+		}
+	}
+	log.Ctx(a.ctx).Info().Str("handler", h.Name).Msg("Shutdown successfull")
+	return nil
+}
+
+// RunAndWait executes the main loop on a go-routine and listens to SIGINT and SIGKILL to start the shutdown
+func (a *App) RunAndWait() error {
+	errs := make(chan error)
+
+	go func() {
+		log.Ctx(a.ctx).Info().Msg("Application main loop starting now!")
+		errs <- a.mainLoop()
+	}()
+
+	signals := make(chan os.Signal, 2)
+	signal.Notify(signals, syscall.SIGINT, syscall.SIGTERM)
+	var err error
+	ctx := log.Ctx(a.ctx).WithContext(context.Background())
+	select {
+	case <-a.ctx.Done():
+		log.Ctx(a.ctx).Info().Msg("App context canceled, initialing shutdown procedures...")
+		err = a.Shutdown(ctx)
+	case s := <-signals:
+		log.Ctx(a.ctx).Info().
+			Str("signal", s.String()).
+			Dur("grace_period", a.GracePeriod).
+			Msg("Signal received. Waiting grace period...")
+		time.Sleep(a.GracePeriod)
```

Comment:
Esse Sleep faz sentido? Quando s for ativado, teve alguma alteração que faz esse Sleep esperar por algo?

Ou a intenção é fazer o programa não entrar em greaceful imediato?

Reply:
Exatamente isso. Grace preriod é um período que a aplicação tem para continuar rodando antes de iniciar o shutdown. Isso permite um fluxo mais de shutdown mais complexo, como tirar uma aplicação de um load balancer, mas permitir ela continuar trabahando por um tempo para que requisições que por algum motivo ainda cheguem (cache do LB) não sejam recusadas. Não sei o quanto isso é util de verdade, mas sempre tivemos isso nos nossos códigos.